### ISSUE TITLE
Face recognition code changes

### ISSUE THREAD

**Issue by gkaretka** [prevso1088/open-set-biometrics#2]
Hi, I would like to change the code to use it for face recognition. I made the following changes since I want to use cosine similarity:

```
def compute_distance(x: torch.Tensor, y: torch.Tensor):
    """
        x: [p, n_x, c]
        y: [p, n_y, c]
    """    
    # Compute cosine similarity
    # Remove first dim, L2 normalize, matmul, add extra dim
    x = x.squeeze(0)
    y = y.squeeze(0)
    x_normed = torch.nn.functional.normalize(x, dim=1)
    y_normed = torch.nn.functional.normalize(y, dim=1)

    result = torch.matmul(x_normed, y_normed.T)
    return result[None, :, :]
```

and implemented it in my code as follow:

```
        self.open_loss = OpenSetLoss(num_parts=1) # since I have 768 dimensional vectors
```

and in the training step I do this:

```
def perform_loss(self, embeddings, labels):    
        # extend to match [n, c, p]
        embeddings = embeddings[:, :, None]

        # Parameters
        S1 = min(8, embeddings.shape[0])  # Number of subjects to select
        S2 = min(4, embeddings.shape[0])  # Number of sequences per subject to select

        # Create a list of unique subjects
        unique_subjects = list(set(labels))

        if len(unique_subjects) == 1:
            self.log_dict({
                'problem': float(-1)
            })
            return 0
        
        # Select S1 subjects without replacement
        selected_subjects = random.sample(unique_subjects, min(S1, len(unique_subjects)))

        # Initialize lists to store selected sequences
        selected_sequences = []

        # Iterate over selected subjects and select S2 sequences per subject with replacement
        for subject in selected_subjects:
            subject_sequences = [i for i, label in enumerate(labels) if label == subject]
            selected_sequences.extend(random.choices(subject_sequences, k=S2))

        # Select the corresponding features, labels, cams, and time_seqs based on selected sequences
        selected_features = embeddings[selected_sequences]
        selected_labels = [unique_subjects.index(labels[i]) for i in selected_sequences]
        embeddings, labels = torch.as_tensor(selected_features), torch.as_tensor(selected_labels)

        return self.open_loss(embeddings, labels)
```

I suspect I did something incorrectly since I get worse performance than just by using plain `arcface` loss. I combine losses as follow:

```
loss = self.loss(y_p, y) + self.perform_loss(y_p, y)
```

where `self.loss` is `arcface`. Thanks

**Comment by prevso1088** [comment]
Hi, thank you so much for your interest in our work!

Based on the code snippets that you shared, we've noticed the following:
- It seems that you only use a subset of each batch for calculating the open-set losses. For data efficiency, we implemented the subject and images/sequences selection procedure in our dataloader and used the entire batch to calculate the losses.
- You used a quite small batch size. In our face recognition experiments. We used S1=512, S2=2.

Hope this helps!

However, we'd also like to note that depending on the setup, the face recognition performance gain is somewhat mixed with the loss in our experiments.

**Comment by gkaretka** [comment]
Hi, thank you for your reply.

- I rewrote my data loaders so that they use the whole batch.
- I also decided to use Euclidean distance instead of cosine similarity.
- I tried using a smaller model to use a larger batch. I notice some nice improvements right away.

Thank you!
